### ISSUE TITLE
stepfunctions: validation on state machine name breaks references to names of other constructs

### ISSUE THREAD

**Issue by sthm** [aws/aws-cdk#13610]
PR #13387 has introduced additional validation on state machine names. However, this validation breaks the capabilities to specify names that are dependent on other constructs or pseudo parameters.

```
new sfn.StateMachine(this, '...', {
            definition: ...,
            stateMachineName: cdk.Aws.STACK_NAME
//          stateMachineName: kafka.clusterName   //also broken
});
```

This definition was working with cdk version 1.92.0 and prior, but now causes the following error.

```
Error: State Machine name must match "^[0-9a-zA-Z+!@._-]+$". Received: ${Token[AWS.StackName.12]}
```

Can you please investigate.

Thanks, Steffen

### Environment

  - **CDK CLI Version:** 1.93.0 (build c506d3b)
  - **Module Version:** 1.93.0
  - **Node.js Version:** v15.11.0
  - **OS:** macOS Catalina
  - **Language (Version):** TypeScript (3.8.3)

### Other information

```
> cdk deploy --require-approval never
Error: State Machine name must match "^[0-9a-zA-Z+!@._-]+$". Received: ${Token[AWS.StackName.12]}
    at StateMachine.validateStateMachineName (/Users/shausma/Code/AwsSaMskKdsPerformance/cdk/node_modules/@aws-cdk/aws-stepfunctions/lib/state-machine.ts:438:13)
    at new StateMachine (/Users/shausma/Code/AwsSaMskKdsPerformance/cdk/node_modules/@aws-cdk/aws-stepfunctions/lib/state-machine.ts:383:12)
    at new CreditDepletion (/Users/shausma/Code/AwsSaMskKdsPerformance/cdk/lib/credit-depletion-sfn.ts:139:29)
    at new CdkStack (/Users/shausma/Code/AwsSaMskKdsPerformance/cdk/lib/cdk-stack.ts:241:37)
    at Object.<anonymous> (/Users/shausma/Code/AwsSaMskKdsPerformance/cdk/bin/cdk.ts:36:1)
    at Module._compile (node:internal/modules/cjs/loader:1092:14)
    at Module.m._compile (/Users/shausma/Code/AwsSaMskKdsPerformance/cdk/node_modules/ts-node/src/index.ts:1056:23)
    at Module._extensions..js (node:internal/modules/cjs/loader:1121:10)
    at Object.require.extensions.<computed> [as .ts] (/Users/shausma/Code/AwsSaMskKdsPerformance/cdk/node_modules/ts-node/src/index.ts:1059:12)
    at Module.load (node:internal/modules/cjs/loader:972:32)
```

**Comment by jstrese** [comment]
@shivlaks This is resolved in PR #13970

**Comment by shivlaks** [comment]
@jstrese thanks for contributing the fix! i'll resolve this issue.
